Code/Tools/FBuild/FBuildCore/Graph/DependencyListNode.cpp

```diff
@@ -0,0 +1,243 @@
+// DependencyListNode.cpp
+//------------------------------------------------------------------------------
+
+// Includes
+//------------------------------------------------------------------------------
+#include "DependencyListNode.h"
+
+#include "Tools/FBuild/FBuildCore/FBuild.h"
+#include "Tools/FBuild/FBuildCore/FLog.h"
+#include "Tools/FBuild/FBuildCore/BFF/Functions/Function.h"
+#include "Tools/FBuild/FBuildCore/Graph/AliasNode.h"
+#include "Tools/FBuild/FBuildCore/Graph/NodeGraph.h"
+
+#include "Core/Containers/Array.h"
+#include "Core/Env/ErrorFormat.h"
+#include "Core/FileIO/FileIO.h"
+#include "Core/FileIO/FileStream.h"
+#include "Core/FileIO/PathUtils.h"
+#include "Core/Strings/AStackString.h"
+
+// REFLECTION
+//------------------------------------------------------------------------------
+REFLECT_NODE_BEGIN( DependencyListNode, Node, MetaNone() )
+    REFLECT(        m_Source,                   "Source",                   MetaFile() )
+    REFLECT(        m_Dest,                     "Dest",                     MetaPath() )
+    REFLECT_ARRAY(  m_Patterns,                 "Patterns",                 MetaOptional() )
+    REFLECT_ARRAY(  m_PreBuildDependencyNames,  "PreBuildDependencies",     MetaOptional() + MetaFile() + MetaAllowNonFile() )
+REFLECT_END( DependencyListNode )
+
+// Anonymous DependencyListBuilder struct
+//------------------------------------------------------------------------------
+struct DependencyListBuilder {
+    enum {
+        iBuckets = 256,
+        mBuckets = (iBuckets - 1)
+    };
+
+    using bucket_t = Array< const Node * >;
+    bucket_t Visiteds[ iBuckets ];
+
+    const Array< AString > & Patterns;
+
+    explicit DependencyListBuilder( const Array< AString > & patterns )
+        : Patterns( patterns )
+    {
+    }
+
+    DependencyListBuilder( const DependencyListBuilder& ) = delete;
+    DependencyListBuilder& operator =( const DependencyListBuilder& ) = delete;
+
+    bool FindOrAddNode( const Node * node )
+    {
+        ASSERT( node );
+
+        bucket_t & bucket = Visiteds[ node->GetNameCRC() & mBuckets ];
+
+        if ( bucket.Find( node ) == nullptr )
+        {
+            bucket.Append( node );
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+
+    void Collect( const Node * node )
+    {
```

Comment:
This function is recursive, but that doesn't seem correct.

If you compile a cpp file, and that depends on a header, that header may be generated. It's pretty common for code-gen to output many headers, so while the upstream dependency of the header might change, the header itself might not. By making the dependencies recursive, you will cause unnecessary rebuilds.